### PR TITLE
PR for issue 28

### PR DESCRIPTION
**Pull request by mpibpc-mroose**
I wrote a proof of concept for implementing responsive tables for Bootstrap v4 and v5. Basically the goal was to use another base template from `django-tables2`. To make it DRY I needed to introduce some partials and move some of the code out of the templates for re-use. 

Roughly tested it in my projects and it completely does fulfill my needs.

### CODE REVIEW

File: django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html [hunk 1]
```diff
@@ -0,0 +1,80 @@
+{% load trans from i18n %}
+{% load static %}
+
+
+
+
```
Comment (djk2) [inline django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html:6]: Hi. 
Can you remove blank lines?

File: django_tables2_column_shifter/templates/django_tables2_column_shifter/bootstrap4-responsive.html [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{% extends "django_tables2/bootstrap4-responsive.html" %}
+
+{% block table %}
+{% if table.shift_table_column %}
+    <div id="{{ table.uniq_table_class_name }}" class="column-shifter-container">
+        {% include 'django_tables2_column_shifter/_partials/bootstrap4_table_block.html' %}
+
+        {# Wrapper default is hide #}
+        <div class="table-wrapper" style="display:none;">
+            {# Load default table content   #}
```
Comment (djk2) [inline django_tables2_column_shifter/templates/django_tables2_column_shifter/bootstrap4-responsive.html:10]:
```suggestion
            {# Load default table content #}
```

File: django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html [hunk 3]
```diff
@@ -0,0 +1,80 @@
+{% load trans from i18n %}
+{% load static %}
+
```
Comment (djk2) [inline django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html:3]:
```suggestion
```

File: django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
```
Comment (djk2) [inline django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html:4]:
```suggestion
```

File: django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
```
Comment (djk2) [inline django_tables2_column_shifter/templates/django_tables2_column_shifter/_partials/bootstrap4_table_block.html:5]:
```suggestion
```

**Review by djk2 — CHANGES_REQUESTED** [review]

**Comment by djk2** [comment]
I have also one general request. 
In repository exist also the `testproject` where you can in quick way launch demo project to test how library works.
Can you also extend this demo project to show how lib works in "responsive" mode. 
In `testproject` on "index/home" page user can choice a version of bootstrap from 2 to 5.  (blue buttons), 
I suggest add 2 more buttons for `boostrap4-responsive` and `boostrap5-responsive`
What do you think ? 

image

On this occasion you can also change a button layout from horizontal to vertical. 
You know :) when I started working on this library I had only 2 versions (buttons) of bootstrap. 
Now there is a five buttons - it will be better to display them in vertical stack.

**Comment by djk2** [comment]
Can you also increase the version of library in `django_tables2_column_shifter/__init__.py` from `2.1.1` to `2.2.0` and add a note in `CHANGELOG.rst` describing changes for this version. ?

**Comment by mpibpc-mroose** [comment]
Should be all resolved now.

**Comment by djk2** [comment]
Can you fix the tests? 
To run test locally you can execute command 
`tox -r --workdir /tmp/tox-dir`
At now, tests returning error: 
```
django.template.exceptions.TemplateDoesNotExist: django_tables2/bootstrap4-responsive.html   
```

To code I do not have more comments.

**Comment by mpibpc-mroose** [comment]
That's pretty interesting. I'm relying on a template from `django-datatables2` which seemed to have introduced in the 2.5 versions (`django_tables2/bootstrap{4,5}-responsive.html`).  I wasn't aware of this. And it makes it pretty difficult for me to decide how to proceed as I was building anything around that.

I'm pretty unsure how to proceed. Any advice?

**Comment by djk2** [comment]
Give me some time to consider how solve it. 
I will try to look on that tomorrow

File: django_tables2_column_shifter/tests/test_base.py [hunk 6]
```diff
@@ -45,13 +47,27 @@ class DjangoTables2ColumnShifterTest(TestCase):
...
         {
             'bootstrap_version': 'bootstrap5',
             'min_dt_version': (2, 0),
             'max_dt_version': None,
             'template_name': 'django_tables2_column_shifter/bootstrap5.html',
             'table_clsss': ColumnShiftTableBootstrap5,
         },
+        {
+            'bootstrap_version': 'bootstrap5responsive',
+            'min_dt_version': (2, 0),
```
Comment (djk2) [inline django_tables2_column_shifter/tests/test_base.py]:
```suggestion
            'min_dt_version': (2, 5, 3),
```

File: django_tables2_column_shifter/tests/test_base.py [hunk 7]
```diff
@@ -45,13 +47,27 @@ class DjangoTables2ColumnShifterTest(TestCase):
...
+        {
+            'bootstrap_version': 'bootstrap4responsive',
+            'min_dt_version': (2, 0),
+            'max_dt_version': None,
+            'template_name': 'django_tables2_column_shifter/bootstrap4-responsive.html',
+            'table_clsss': ColumnShiftTableBootstrap4Responsive,
+        },
         {
             'bootstrap_version': 'bootstrap5',
             'min_dt_version': (2, 0),
```
Comment (djk2) [inline django_tables2_column_shifter/tests/test_base.py]:
```suggestion
            'min_dt_version': (2, 5),
```

File: django_tables2_column_shifter/tests/test_base.py [hunk 8]
```diff
@@ -45,13 +47,27 @@ class DjangoTables2ColumnShifterTest(TestCase):
             'template_name': 'django_tables2_column_shifter/bootstrap4.html',
             'table_clsss': ColumnShiftTableBootstrap4,
         },
+        {
+            'bootstrap_version': 'bootstrap4responsive',
+            'min_dt_version': (2, 0),
```
Comment (djk2) [inline django_tables2_column_shifter/tests/test_base.py]:
```suggestion
            'min_dt_version': (2, 5),
```

File: CHANGELOG.rst [hunk 9]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 ===========
 
+v. 2.2.0
+--------
+
+    * Add new classes ``ColumnShiftTableBootstrap4Responsive`` and ``ColumnShiftTableBootstrap5Responsive``
```
Comment (djk2) [inline CHANGELOG.rst]: I suggest redraft: 

```
    * Add new classes:
        * ``ColumnShiftTableBootstrap4Responsive`` - only for django-tables2 >= 2.5.0          
        * ``ColumnShiftTableBootstrap5Responsive`` - only for django-tables2 >= 2.5.3
    
      Classes able to use responsive tables with Bootstrap v4 and v5 
      Author: @mpibpc-mroose
```

File: README.rst [hunk 10]
```diff
@@ -33,8 +33,8 @@ Using JQuery, Bootstrap3 or Bootstrap4 or Bootstrap5 and Django >=1.9.
 
   * for bootstrap2 - ColumnShiftTableBootstrap2,
   * for bootstrap3 - ColumnShiftTableBootstrap3,
-  * for bootstrap4 - ColumnShiftTableBootstrap4,
-  * for bootstrap5 - ColumnShiftTableBootstrap5,
+  * for bootstrap4 - ColumnShiftTableBootstrap4 or ColumnShiftTableBootstrap4Responsive,
+  * for bootstrap5 - ColumnShiftTableBootstrap5 or ColumnShiftTableBootstrap5Responsive.
```
Comment (djk2) [inline README.rst]: Redraft to: 
```
  * for bootstrap4 - ColumnShiftTableBootstrap4 
  * for bootstrap4 in responsive mode - ColumnShiftTableBootstrap4Responsive (only for django-tables2 >= 2.5)
  * for bootstrap5 - ColumnShiftTableBootstrap5 
  * for bootstrap5 in responsive mode - ColumnShiftTableBootstrap5Responsive (only for django-tables2 >= 2.5.3)
```

File: README.rst [hunk 11]
```diff
@@ -218,11 +218,16 @@ Bootstrap4 :
 If you use Bootstrap v4 in your project then table class has to inherit from `ColumnShiftTableBootstrap4`
 imported from `django_tables2_column_shifter.tables`.
 
+Alternatively if you want to use `table-responsive` your table class has to inherit from
```
Comment (djk2) [inline README.rst:223]: add here that responsive mode is available for django-tables2 v2.5 and 2.5.3 only 
I want to be clear for everyone that this feature if available only in newest version of django-tables2

File: README.rst [hunk 12]
```diff
@@ -218,11 +218,16 @@ Bootstrap4 :
...
+`ColumnShiftTableBootstrap4Responsive` imported from `django_tables2_column_shifter.tables`.
+
 Bootstrap5:
 --------------------------------------
 If you use Bootstrap v5 in your project then table class has to inherit from `ColumnShiftTableBootstrap5`
 imported from `django_tables2_column_shifter.tables`.
 
+Alternatively if you want to use `table-responsive` your table class has to inherit from
```
Comment (djk2) [inline README.rst:231]: As above, add here info that table-responsive is available only for correct version of django-tables2

File: django_tables2_column_shifter/tests/test_base.py [hunk 13]
```diff
@@ -12,7 +12,9 @@
     ColumnShiftTableBootstrap2,
```
Comment (djk2) [inline django_tables2_column_shifter/tests/test_base.py:13]:
```suggestion
    dt_version,
    ColumnShiftTableBootstrap2,
```

File: django_tables2_column_shifter/tests/test_base.py [hunk 14]
```diff
@@ -45,13 +47,27 @@ class DjangoTables2ColumnShifterTest(TestCase):
...
+        {
+            'bootstrap_version': 'bootstrap5responsive',
+            'min_dt_version': (2, 0),
+            'max_dt_version': None,
+            'template_name': 'django_tables2_column_shifter/bootstrap5-responsive.html',
+            'table_clsss': ColumnShiftTableBootstrap5Responsive,
+        },
     ]
 
     dt_version = tuple(map(int, tables.__version__.split(".")[:2]))
```
Comment (djk2) [inline django_tables2_column_shifter/tests/test_base.py]:
```suggestion
```
And you have to replace in whole file a `self.dt_version` => `dt_version`  now you are importing it from `django_tables2_column_shifter.tables`

File: django_tables2_column_shifter/tables.py [hunk 15]
```diff
@@ -2,7 +2,6 @@
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py:2]:
```suggestion

dt_version = tuple(map(int, tables.__version__.split(".")[:3]))

```

File: django_tables2_column_shifter/tables.py [hunk 16]
```diff
@@ -89,9 +88,23 @@ class ColumnShiftTableBootstrap4(ColumnShiftTable):
...
+    shifter_template = "django_tables2_column_shifter/bootstrap4-responsive.html"
+
+
 class ColumnShiftTableBootstrap5(ColumnShiftTable):
     """
     Table class compatible with bootstrap 5
     """
     dropdown_button_css = "btn btn-light btn-sm"
     shifter_template = "django_tables2_column_shifter/bootstrap5.html"
+
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py:114]:
```suggestion

    def __init__(self, *args, **kwargs):
        if dt_version < (2, 5):
            raise AssertionError(
                "ColumnShiftTableBootstrap5 require django-tables2 >= 2.5 "
                "your current version is {}".format(tables.__version__)
            )
        super(ColumnShiftTableBootstrap5, self).__init__(*args, **kwargs)
```

File: django_tables2_column_shifter/tables.py [hunk 17]
```diff
@@ -89,9 +88,23 @@ class ColumnShiftTableBootstrap4(ColumnShiftTable):
...
+
+class ColumnShiftTableBootstrap5Responsive(ColumnShiftTableBootstrap5):
+    """
+    Table class compatible with Bootstrap 5 and using "table-responsive" css class.
+    """
+    shifter_template = "django_tables2_column_shifter/bootstrap5-responsive.html"
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py:128]:
```suggestion
    shifter_template = "django_tables2_column_shifter/bootstrap5-responsive.html"
    
    def __init__(self, *args, **kwargs):
        if dt_version < (2, 5, 3):
            raise AssertionError(
                "ColumnShiftTableBootstrap5Responsive require django-tables2 >= 2.5.3 "
                "your current version is {}".format(tables.__version__)
            )
        super(ColumnShiftTableBootstrap5Responsive, self).__init__(*args, **kwargs)

```

File: django_tables2_column_shifter/tables.py [hunk 18]
```diff
@@ -89,9 +88,23 @@ class ColumnShiftTableBootstrap4(ColumnShiftTable):
     shifter_template = "django_tables2_column_shifter/bootstrap4.html"
 
 
+class ColumnShiftTableBootstrap4Responsive(ColumnShiftTable):
+    """
+    Table class compatible with Bootstrap 4 and using "table-responsive" css class.
+    """
+    shifter_template = "django_tables2_column_shifter/bootstrap4-responsive.html"
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py:97]:
```suggestion
    shifter_template = "django_tables2_column_shifter/bootstrap4-responsive.html"
    
    def __init__(self, *args, **kwargs):
        if dt_version < (2, 5):
            raise AssertionError(
                "ColumnShiftTableBootstrap4Responsive require django-tables2 >= 2.5 "
                "your current version is {}".format(tables.__version__)
            )
        super(ColumnShiftTableBootstrap4Responsive, self).__init__(*args, **kwargs)

```

**Review by djk2 — CHANGES_REQUESTED** [review]

**Comment by djk2** [comment]
> That's pretty interesting. I'm relying on a template from `django-datatables2` which seemed to have introduced in the 2.5 versions (`django_tables2/bootstrap{4,5}-responsive.html`). I wasn't aware of this. And it makes it pretty difficult for me to decide how to proceed as I was building anything around that.
> 
> I'm pretty unsure how to proceed. Any advice?

Hi. 
I added suggestions to code. 
Changes are not vast. 
I proposed to add a exclusions and inform users that they can use new features only for some of versions of django-tables2. 
I included changes in tests. 
You have also fix the pep8. 
```
./django_tables2_column_shifter/tests/urls.py:3:101: E501 line too long (109 > 100 characters)
./testproject/testproject/urls.py:21:101: E501 line too long (102 > 100 characters)
./testproject/testproject/urls.py:24:101: E501 line too long (102 > 100 characters)
```

File: django_tables2_column_shifter/tests/test_base.py (hunk 13 above)
Comment (djk2) [reply django_tables2_column_shifter/tests/test_base.py:13]: Please import `dt_version` from `django_tables2_column_shifter.tables` you added it there.

File: django_tables2_column_shifter/tests/test_base.py [hunk 19]
```diff
@@ -45,13 +47,27 @@ class DjangoTables2ColumnShifterTest(TestCase):
...
+        {
+            'bootstrap_version': 'bootstrap5responsive',
+            'min_dt_version': (2, 5),
+            'max_dt_version': None,
+            'template_name': 'django_tables2_column_shifter/bootstrap5-responsive.html',
+            'table_clsss': ColumnShiftTableBootstrap5Responsive,
+        },
     ]
 
     dt_version = tuple(map(int, tables.__version__.split(".")[:2]))
```
Comment (djk2) [inline django_tables2_column_shifter/tests/test_base.py]: Can you remove `dt_version` from here and import it from  `django_tables2_column_shifter.tables ` at top of file. 
After yesterday changes we already have `dt_version` in `tables`
Moreover, now you need version containing 3 digits instead 2.

**Review by djk2 — CHANGES_REQUESTED** [review]

File: django_tables2_column_shifter/tests/test_base.py [hunk 20]
```diff
@@ -12,8 +12,11 @@
     ColumnShiftTableBootstrap2,
     ColumnShiftTableBootstrap3,
     ColumnShiftTableBootstrap4,
+    ColumnShiftTableBootstrap4Responsive,
     ColumnShiftTableBootstrap5,
+    ColumnShiftTableBootstrap5Responsive
 )
+from django_tables2_column_shifter.tables import dt_version
```
Comment (djk2) [inline django_tables2_column_shifter/tests/test_base.py]: Put  this import between 11 and 12 line. 
Above package `tables` is already imported from `django_tables2_column_shifter`

Comment (mpibpc-mroose) [reply django_tables2_column_shifter/tests/test_base.py]: Your absolutely right!

**Pull request merged**